File: leetcode/medium/208-implementTriePrefixTree.js
// Implement a trie with insert, search, and startsWith methods.

// Example:

// Trie trie = new Trie();

// trie.insert("apple");
// trie.search("apple");   // returns true
// trie.search("app");     // returns false
// trie.startsWith("app"); // returns true
// trie.insert("app");   
// trie.search("app");     // returns true

// first pass solution using an ES6 class and set for data storage
class Trie {
    constructor() {
        this.trie = new Set();
    }
    
    insert(word) {
        if (!(this.trie.has(word))) {
            this.trie.add(word);
        }
    }
    
    search(word) {
        if (this.trie.has(word)) {
            return true;
        }
        return false;
    }
    
    startsWith(prefix) {
        for (let word of this.trie) {
            if (word.startsWith(prefix)) {
                return true;
            }
        }
        return false;
    }
}

File: leetcode/medium/208-implementTriePrefixTree.java
// first pass solution
class TrieNode {
    Map<Character, TrieNode> children = new HashMap<>();
    boolean isEnd = false;
}

class Trie {
    TrieNode root = new TrieNode();
    
    /** Initialize your data structure here. */
    public Trie() {}
    
    /** Inserts a word into the trie. */
    public void insert(String word) {
        TrieNode node = root;
        
        for (char currChar : word.toCharArray()) {
            if (!node.children.containsKey(currChar)) {
                node.children.put(currChar, new TrieNode());
            }
            node = node.children.get(currChar);
        }
        node.isEnd = true;
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        TrieNode node = root;
        
        for (char currChar : word.toCharArray()) {
            if (node.children.get(currChar) == null) {
                return false;
            }
            node = node.children.get(currChar);
        }
        return node.isEnd;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        TrieNode node = root;
        
        for (char currChar : prefix.toCharArray()) {
            if (!node.children.containsKey(currChar)) {
                return false;
            }
            node = node.children.get(currChar);
        }
        return true;
    } 
}

// second pass solution
class TrieNode {
    TrieNode[] children = new TrieNode[26];
    boolean isEnd = false;
}

class Trie {
    TrieNode root = new TrieNode();
    
    /** Initialize your data structure here. */
    public Trie() {}
    
    /** Inserts a word into the trie. */
    public void insert(String word) {
        TrieNode node = root;
        
        for (char c : word.toCharArray()) {
            int index = (int) c -'a';
            if (node.children[index] == null) {
                node.children[index] = new TrieNode();
            }
            node = node.children[index];
        }
        node.isEnd = true;
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        TrieNode node = root;
        
        for (char c : word.toCharArray()) {
            int index = (int) c -'a';
            if (node.children[index] == null) {
                return false;
            }
            node = node.children[index];
        }
        return node.isEnd;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        TrieNode node = root;
        
        for (char c : prefix.toCharArray()) {
            int index = (int) c -'a';
            if (node.children[index] == null) {
                return false;
            }
            node = node.children[index];
        }
        return true;
    } 
}
